lib/handler/cache-handler.js

```diff
@@ -239,15 +248,30 @@ class CacheHandler {
  * @param {number} statusCode
  * @param {import('../../types/header.d.ts').IncomingHttpHeaders} resHeaders
  * @param {import('../../types/cache-interceptor.d.ts').default.CacheControlDirectives} cacheControlDirectives
+ * @param {number | undefined} cacheByDefault
  */
-function canCacheResponse (cacheType, statusCode, resHeaders, cacheControlDirectives) {
-  // Allow caching for status codes 200 and 307 (original behavior)
-  // Also allow caching for other status codes that are heuristically cacheable
-  // when they have explicit cache directives
-  if (statusCode !== 200 && statusCode !== 307 && !HEURISTICALLY_CACHEABLE_STATUS_CODES.includes(statusCode)) {
+function canCacheResponse (cacheType, statusCode, resHeaders, cacheControlDirectives, cacheByDefault) {
+  // Status code must be final and understood.
+  if (statusCode < 200 || NOT_UNDERSTOOD_STATUS_CODES.includes(statusCode)) {
     return false
   }
 
+  if (cacheByDefault) {
+    // When cacheByDefault is truthy, the cache interceptor does no more mandate an explicit expiry prior to
+    // reaching here. We should then check the full condition allowing to consider caching responses.
+    // Responses with neither status codes that are heuristically cacheable, nor "explicit enough" caching
+    // directives, are not cacheable. "Explicit enough": see https://www.rfc-editor.org/rfc/rfc9111.html#section-3
+    if (!HEURISTICALLY_CACHEABLE_STATUS_CODES.includes(statusCode) && !resHeaders['expires'] &&
+      !cacheControlDirectives.public &&
+      cacheControlDirectives['max-age'] === undefined &&
+      // RFC 9111: a private response directive, if the cache is not shared
+      !(cacheControlDirectives.private && cacheType === 'private') &&
+      !(cacheControlDirectives['s-maxage'] !== undefined && cacheType === 'shared')
+    ) {
+      return false
+    }
+  }
```

Comment:
When I considered the subject, I was not expecting it to turn that way because I was not aware of the `cacheByDefault` option.

Do to the code preceding the call to canCacheResponse, which actually does already some cacheability checks, we need that big test only when cacheByDefault is truthy.

See https://github.com/nodejs/undici/blob/4608ef157cc75d5ce3d2802e6949cb865d73146c/lib/handler/cache-handler.js#L115C1-L125C6

Maybe that code should be moved into `canCacheResponse` instead of staying out of it, allowing to have a better picture of how the cache interceptor check cacheability.